Spurious "Expression has type Any" error following attribute access

As originally described in #5788, I've run into a problem with `--disallow-any-expr` reporting spurious errors.

Consider the following file `any.py`:

```python3
def access_before_declaration(self) -> None:
    obj = Object('foo')
    obj.value
    x = 1

    reveal_type(x) # Revealed type is 'builtins.int'
    x = x + 1 # Expression has type "Any"

class Object:
    def __init__(self, value: str) -> None:
        self.value = value

def access_after_declaration(self) -> None:
    obj = Object('foo')
    obj.value
    x = 1

    reveal_type(x) # Revealed type is 'builtins.int'
    x = x + 1
```

Running mypy on it produces:

```
$ python3 -m mypy --version
mypy 0.641
$ python3 -m mypy --disallow-any-expr any.py
any.py:6: error: Revealed type is 'builtins.int'
any.py:7: error: Expression has type "Any"
any.py:18: error: Revealed type is 'builtins.int'
```

What we see is that `x` is correctly identified as `int` in `access_after_declaration`, but incorrectly as `Any` in `access_before_declaration`. The error is only triggered in very specific circumstances:

* There is a class defined **after** the inspected line
* An attribute of an instance of that class is accessed **before** the inspected line

Just a small update, this can still be reproduced on the latest mypy:

```
$ python3 -m mypy --version
mypy 0.740
$ python3 -m mypy --disallow-any-expr any.py
any.py:6: error: Expression has type "Any"
any.py:6: note: Revealed type is 'builtins.int'
any.py:7: error: Expression has type "Any"
any.py:18: note: Revealed type is 'builtins.int'
Found 2 errors in 1 file (checked 1 source file)
```

(*Almost* the same result as before, now with an additional `Any` error for `reveal_type(x)` in line 6)

FYI, we use `--disallow-any-expr` for our now completely annotated code base, and this bug triggers dozens of errors. Is there anything we can do to move this issue forward? :)

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -3613,7 +3613,7 @@ class ExpressionChecker(ExpressionVisitor[Type]):
                 not always_allow_any and
                 not self.chk.is_stub and
                 self.chk.in_checked_function() and
-                has_any_type(typ)):
+                has_any_type(typ) and not self.chk.current_node_deferred):
             self.msg.disallowed_any_type(typ, node)
 
         if not self.chk.in_checked_function() or self.chk.current_node_deferred:

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -6489,3 +6489,26 @@ reveal_type(Foo().y)  # N: Revealed type is 'builtins.list[Any]'
 class Foo:
     def bad():  # E: Method must have at least one argument
         self.x = 0  # E: Name 'self' is not defined
+
+[case testTypeAfterAttributeAccessWithDisallowAnyExpr]
+# flags: --disallow-any-expr
+
+def access_before_declaration(self) -> None:
+    obj = Foo('bar')
+    obj.value
+    x = 1
+
+    reveal_type(x)  # N: Revealed type is 'builtins.int'
+    x = x + 1
+
+class Foo:
+    def __init__(self, value: str) -> None:
+        self.value = value
+
+def access_after_declaration(self) -> None:
+    obj = Foo('bar')
+    obj.value
+    x = 1
+
+    reveal_type(x)  # N: Revealed type is 'builtins.int'
+    x = x + 1

